Clear cache after a scheduler update is queued

zaerl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes: https://github.com/Automattic/wp-calypso/issues/90627
Context: p1715613932876249-slack-C01A60HCGUA

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Force a cache clear the first ten seconds after a schedule creation, so that scheduled events at the same second will obtain a fresh cache and not previous value
* `jetpack_scheduled_update_updated` is run only on successful scheduled event creation

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Add this branch to Jetpack Beta
* Quickly add/create/modify a scheduled update. Better when the clock is near second zero (08:00:57 -> 08:01:03, etc.)
* Ensure everything is ok

zaerl committed:
Add additional check after scheduled update creation

zaerl committed:
changelog

zaerl committed:
Update: new clear cache strategy

zaerl committed:
Fix: clean strategy

zaerl committed:
Additional cron cache reload

zaerl committed:
Revert cache clear

zaerl committed:
New cache clear technique

zaerl committed:
Add cache lock

zaerl committed:
Add new clear cache

obenland commented on projects/packages/scheduled-updates/src/class-scheduled-updates.php:
```diff
@@ -91,6 +91,8 @@ public static function init() {
 		add_action( 'add_option_' . Scheduled_Updates_Active::OPTION_NAME, $callback );
 		add_action( 'update_option_' . Scheduled_Updates_Active::OPTION_NAME, $callback );
+
+		add_filter( 'pre_schedule_event', array( __CLASS__, 'clear_cron_cache_pre' ) );
```
Where does the filter get applied?

zaerl replied:
Inside the `wp_schedule_event` function.

obenland replied:
Gotcha, thanks. In the callback, should we check that the event hook is our hook before clearing caches? Or does it need to run for every event scheduling?

zaerl replied:
No because the problem is with the other events that obtain the wrong cacye, not ours. It will run for the first event that will be queued the first 10 seconds after our.

obenland commented on projects/packages/scheduled-updates/src/class-scheduled-updates.php:
```diff
@@ -169,10 +171,15 @@ public static function run_scheduled_update( ...$plugins ) {
 	 */
 	public static function create_scheduled_update( $timestamp, $interval, $plugins ) {
-		return wp_schedule_event( $timestamp, $interval, self::PLUGIN_CRON_HOOK, $plugins, true );
+		$res = wp_schedule_event( $timestamp, $interval, self::PLUGIN_CRON_HOOK, $plugins, true );
```
What's "res"? Could we find a more descriptive variable name?

zaerl replied:
Changed, thanks.

obenland commented on projects/packages/scheduled-updates/src/class-scheduled-updates.php:
```diff
@@ -226,6 +235,17 @@ function ( $event ) use ( $updated_at ) {
 	 */
 	public static function clear_cron_cache() {
 		wp_cache_delete( 'alloptions', 'options' );
+		wp_cache_delete( 'notoptions', 'options' );
```
Should this function be named `clear_options_cache`?

Can you tell me more about clearing notoptions? I'd expect `cron` to never be part of that, is that not the case?

zaerl replied:
I think this should be named `clear_cron_cache`. Although we are deleting all the cache, this is done because of only the previously applied `cron` option. Deleting just the cron entry does not solve the problems.

obenland replied:
> Deleting just the cron entry does not solve the problems.

Do you know why that is?

zaerl replied:
We had a very long debug session but we cannot find the exact reason why. Probably something related to other plugins running on WoA (like the `action-scheduler`). p1715925157297969-slack-C01A60HCGUA

obenland replied:
Can you tell me more about clearing notoptions? I'd expect cron to never be part of that, is that not the case?

ouikhuan replied:
I think the reason we failed before is probably due to the timing we tried to clear the cache wrong before. We used
```
wp_cache_delete( 'alloptions', 'options' );
wp_cache_delete( 'notoptions', 'options' );
``` 
From looking into the codebase and see what's the proper way of clearing all the cache, and placing it for testing, but I agree if we can only aim for the `cron` option that'll be the best. In our current implementation, I tried to test it out again with only:

```
wp_cache_delete( 'cron', 'options' );
wp_load_alloptions( true );
```

And looks like it works too, @zaerl could you change it back to this and see if it works for you?

zaerl replied:
> And looks like it works too, @zaerl could you change it back to this and see if it works for you?

Confirmed it is working https://github.com/Automattic/jetpack/pull/37399/commits/8ee38f6f60dde28d95d4bf919c211d4249c3fd0d

obenland commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -179,13 +179,21 @@ public function create_item( $request ) {
 
-		$id = Scheduled_Updates::generate_schedule_id( $plugins );
+		$id     = Scheduled_Updates::generate_schedule_id( $plugins );
+		$events = wp_get_scheduled_events( Scheduled_Updates::PLUGIN_CRON_HOOK );
```
We should be able to use `wp_get_scheduled_event()` here instead.

zaerl replied:
Done.

obenland commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -194,12 +202,15 @@ public function create_item( $request ) {
 		$this->update_additional_fields_for_object( $event, $request );
 
+		// Lock the cache after a successful schedule creation.
+		set_transient( 'jetpack_scheduled_update_created_lock', true, 10 );
```
How does it lock the cache?

zaerl replied:
Because if a scheduled event is added at the exact second, it obtains the old cache, and so the cache should be cleared the very first time a new event is scheduled. Despite all the cache clearing done during `update_option` and others.

obenland replied:
With "lock" I'd expect there to be no changes to the cache, but we're deleting it.

zaerl replied:
What name is the best to you for this transient?

ouikhuan replied:
If we're consider moving this into a more general place, perhaps something like `pre_schedule_event_clear_cron_cache` ?

zaerl replied:
Renamed.

obenland commented on projects/packages/scheduled-updates/src/class-scheduled-updates.php:
```diff
@@ -183,10 +190,12 @@ public static function create_scheduled_update( $timestamp, $interval, $plugins
 		self::clear_cron_cache();
 
-		return wp_unschedule_event( $timestamp, self::PLUGIN_CRON_HOOK, $plugins, true );
+		return $ret;
```
Would this be an opportunity to move `wp_unschedule_event()` back into `delete_item()` [and hook ](https://github.com/Automattic/jetpack/pull/37007/files#diff-16bb61901772bfecbc9020e4b205240d10a347c6f5e6ac61fe795186bcc4d997R62)`clear_cron_cache()` into `jetpack_scheduled_update_deleted`?

zaerl replied:
This idea is excellent, but it can be added in a separate PR.

zaerl committed:
Rename variable

zaerl committed:
Change wp_get_scheduled_events to wp_get_scheduled_event

obenland commented on projects/packages/scheduled-updates/src/class-scheduled-updates.php:
```diff
@@ -183,10 +190,12 @@ public static function create_scheduled_update( $timestamp, $interval, $plugins
 	 */
 	public static function delete_scheduled_update( $timestamp, $plugins ) {
-		// Be sure to clear the cron cache before removing a cron entry.
+		$ret = wp_unschedule_event( $timestamp, self::PLUGIN_CRON_HOOK, $plugins, true );
```
Could we find a more descriptive variable name?

zaerl replied:
Renamed.

obenland commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -179,13 +179,21 @@ public function create_item( $request ) {
 		usort( $plugins, 'strnatcasecmp' );
 
-		$event = Scheduled_Updates::create_scheduled_update( $schedule['timestamp'], $schedule['interval'], $plugins );
+		$res = Scheduled_Updates::create_scheduled_update( $schedule['timestamp'], $schedule['interval'], $plugins );
```
Can we leave `$event`? That's what we should expect to get returned here

zaerl replied:
Renamed.

obenland commented:
@ouikhuan @zaerl I don't know if Scheduled Updates is the best place for this, but after reading through the thread (p1715925157297969-slack-C01A60HCGUA), I do wonder if clearing cron option cache on `pre_schedule_event` for all schedules would be the most straight forward way of solving this? We should also see if we can limit it to the `cron` entry only, and if not, _know_ why not.

Based on your testing it looks like updating cron array with stale data is not just a problem with ActionScheduler but with any event that gets scheduled in close timely proximity to one another?

ouikhuan commented:
> @ouikhuan @zaerl I don't know if Scheduled Updates is the best place for this, but after reading through the thread (p1715925157297969-slack-C01A60HCGUA), I do wonder if clearing cron option cache on `pre_schedule_event` for all schedules would be the most straight forward way of solving this?

Yeah, I agree. This seems to be a more general problem that happens whenever two schedule events run at the same second, and might be better to place this fix somewhere else.

> We should also see if we can limit it to the `cron` entry only, and if not, _know_ why not.

I think we could try to limit it to `cron` only, explain [here](https://github.com/Automattic/jetpack/pull/37399#discussion_r1606210184).
 
> Based on your testing it looks like updating cron array with stale data is not just a problem with ActionScheduler but with any event that gets scheduled in close timely proximity to one another?

That's correct. It can also be some other events like `jetpack_sync_full_cron` too.

ouikhuan reviewed (commented):
@zaerl , I think we'll also need to add the "lock" when we try to delete a schedule, since it might also happens when another schedule event is triggered, what do you think?
Here's a scenario:
1. wp_unschedule_event runs -> the cron get removed.
2. At same second another schedule event triggered ( can be jetpack_sync_full_cron )
3. The cron value gets overwritten again.

zaerl committed:
Update option clearing schema

zaerl committed:
Rename transient

zaerl committed:
Update variable name

zaerl committed:
Rename variable

zaerl committed:
Remove *_scheduled_update functions

zaerl committed:
Add new unit tests

zaerl commented:
@obenland @ouikhuan 
1. Removed the `create_scheduled_update/delete_scheduled_update` functions. Not needed anymore.
2. Changed the cache cleaning method.
3. `delete_item` now deletes the cache only once time, so it's not deleted multiple times during the same call.
4. The `pre_schedule_event_clear_cron_cache` now checks if the event differs from the scheduled updates event.
5. Multiple unit tests were added that counted how many times a transient was added.

```
create_item()
    wp_schedule_event > pre_schedule_event_clear_cron_cache
update_item()
    delete_item()
        wp_unschedule_event
    create_item()
        wp_schedule_event > pre_schedule_event_clear_cron_cache
delete_item()
    wp_unschedule_event > pre_schedule_event_clear_cron_cache
```

obenland commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -179,13 +179,21 @@ public function create_item( $request ) {
+		$event = wp_get_scheduled_event( Scheduled_Updates::PLUGIN_CRON_HOOK, $plugins, $schedule['timestamp'] );
+
+		if ( ! $event ) {
+			return new WP_Error( 'rest_invalid_schedule', __( 'The schedule could not be created.', 'jetpack-scheduled-updates' ), array( 'status' => 500 ) );
```
Should this be a 404? It's not a server level error. There doesn't seem to be a response code that's made for this situation

ouikhuan replied:
How about 422? (Unprocessable Content)

obenland replied:
> Warning: The client should not repeat this request without modification.
> —https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422

I think this is just a really tricky situation

zaerl replied:
I've put on 500 because this shouldn't (but unfortunately do) happen when `wp_schedule_event` don't send back a `WP_Error`. You see it in other failing functions like `wp_update_comment`.

A 404 is better here because something has been created but cannot be found later?

ouikhuan replied:
> I think this is just a really tricky situation

Yeah, that's true. In this case, 404 looks more suitable for our situation here.

zaerl replied:
Changed to 404, thanks.

ouikhuan commented:
@obenland where do you think it's the best for us to place this fix? Should we put it on the Scheduled Updates package as an initial and run for a while to make sure we don't bring in any side effects and move it to a more general place later? Or should we move it to a more general place directly? If so, where do you think is better? Jetpack plugin or wpcomsh?

obenland commented:
> where do you think it's the best for us to place this fix?

Yeah, that's a good question.

With the transient it should probably live in scheduled updates. If we removed that transient, it might make sense to put in wpcomsh with review from others. Ultimately it sounds like a consideration for Core?

zaerl committed:
Update status to 404

ouikhuan commented:
> With the transient it should probably live in scheduled updates. If we removed that transient, it might make sense to put in wpcomsh with review from others. Ultimately it sounds like a consideration for Core?

Yes, I think we'll need to have transient so it only clears the cache when it has to. And since this is more like a general issue for the schedule event system I think it'll be nice to consider adding this to Core eventually.

ouikhuan commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -179,13 +179,21 @@ public function create_item( $request ) {
 
-		$id = Scheduled_Updates::generate_schedule_id( $plugins );
+		$id    = Scheduled_Updates::generate_schedule_id( $plugins );
+		$event = wp_get_scheduled_event( Scheduled_Updates::PLUGIN_CRON_HOOK, $plugins, $schedule['timestamp'] );
```
Since this is added here, is the one on line 207 still needed?

zaerl replied:
Not anymore, you're right. Removed.

ouikhuan commented on projects/packages/scheduled-updates/tests/php/class-wpcom-rest-api-v2-endpoint-update-schedules-test.php:
```diff
@@ -796,26 +856,130 @@ public function test_crud_should_sync_only_three_times() {
+		// Insert a full cron sync event.
+		wp_schedule_event( time(), 'jetpack_sync_interval', 'jetpack_sync_full_cron' );
+
+		$request = new WP_REST_Request( 'GET', '/wpcom/v2/update-schedules' );
+		$result  = rest_do_request( $request );
+
+		$this->assertSame( 200, $result->get_status() );
```
The test is to check if the cron gets reset after another schedule event, but I not sure what's the purpose of having a full cron sync event here to test for, am I missing something?

zaerl replied:
With our latest changes, this unit test can be removed.

ouikhuan commented on projects/packages/scheduled-updates/src/class-scheduled-updates.php:
```diff
@@ -225,7 +201,20 @@ function ( $event ) use ( $updated_at ) {
+	 */
+	public static function clear_cron_cache_pre( $event ) {
+		// If the transient is set and an external event is about to run, it means that the cron cache must be refreshed.
+		if ( self::PLUGIN_CRON_HOOK !== $event && get_transient( 'pre_schedule_event_clear_cron_cache' ) ) {
```
According to [the doc](https://developer.wordpress.org/reference/functions/wp_schedule_event/) It seems the first param is result and the second one is $event object.

zaerl replied:
Yes thanks, changed.

ouikhuan commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -179,13 +179,21 @@ public function create_item( $request ) {
 
 		if ( is_wp_error( $event ) ) {
+			// If the schedule could not be created, return an error.
+			$event->add_data( array( 'status' => 500 ) );
```
Should this be 404 too?

zaerl replied:
I set it to 404 for completeness. Thanks.

zaerl committed:
Minor changes

zaerl committed:
Remove unit test

zaerl committed:
Fix: wrong version

zaerl committed:
Fix: static analysis

ouikhuan reviewed (approved):
Tested several times with other schedule events running at the same second and it looks great. Thanks for handling all the feedback!
